### PR TITLE
Fix/remove unkown ips

### LINKED ISSUE #15: Remove unkown IPs from the endpoint
**What would you like to be added**:
The kubernetes endpoint sometimes is configured with outdated or manually added IPs which are not serving the kube-apiserver of the shoot. Because of this, applications deployed in the clusters are sporadically failing to connect to the kube-apiserver if they hit the wrong IP.

The readvertiser observes this behavior, but it only ensures that the currently active IPs are injected into the endpoint. 
```
time="2020-11-06T12:49:02Z" level=info msg="DNS lookup results are: [10.10.10.10 10.10.10.11]"
time="2020-11-06T12:49:02Z" level=info msg="Kubernetes Endpoint IPs : [\"10.20.20.20\" \"10.10.10.10\" \"10.10.10.11\" \"10.20.20.21\" \"10.20.20.22\" \"10.20.20.22\"]"
time="2020-11-06T12:49:02Z" level=info msg="Nothing to be done"
```

**Why is this needed**:
To ensure that the in-cluster applications are always routed to the API server of the cluster.

### PR DESCRIPTION
**Pull request by BeckerMax**
**Which issue(s) this PR fixes**:
Fixes #15

**Special notes for your reviewer**:
- First commit is upgrading to go modules
- Second commit adds tests and updates the client go version. The tests do not find the bug. They just verified that the `applyTwoWayEndpointMergePatch` function was not the problem here and works as expected. Generally I do not like to add tests for private functions but it seems it adds value here as a documentation how the function works.
- Third commit fixes the problem. It can also be applied without the other commits.

I have two questions to the reviewers:
- Is ist ok to update the k8s.io/client-go to v0.18.2
- I am unsure how to correctly use the context. Should I just pass in the context I received in the Run() method?

**How I tested it**
Added an additional ip to the ep (1.2.3.4)

_log from current version:_
: time="2020-11-11T10:07:55Z" level=info msg="DNS lookup results are: [3.124.60.113 3.125.249.32]"
time="2020-11-11T10:07:55Z" level=info msg="Kubernetes Endpoint IPs : [\"1.2.3.4\" \"3.124.60.113\" \"3.125.249.32\"]"
time="2020-11-11T10:07:55Z" level=info msg="Nothing to be done"

_log with this fix:_
INFO[0005] Kubernetes Endpoint IPs : ["1.2.3.4" "3.124.60.113" "3.125.249.32"]
INFO[0005] ELB records changed, reconciling cluster endpoint to match
INFO[0005] Old endpoint IPs are ["1.2.3.4" "3.124.60.113" "3.125.249.32"], new endpoint IPs are ["3.124.60.113" "3.125.249.32"], ELB IPs are [3.124.60.113 3.125.249.32]
INFO[0010] DNS lookup results are: [3.124.60.113 3.125.249.32]
INFO[0010] Kubernetes Endpoint IPs : ["3.124.60.113" "3.125.249.32"]
INFO[0010] Nothing to be done

```improvement operator
Fix a bug which could cause occasional unavailability of shoot api servers on AWS
```

### CODE REVIEW

**Comment by BeckerMax** [comment]
cc @vpnachev @rfranzke

File: controller/endpoints.go [hunk 1]
```diff
@@ -80,7 +80,7 @@ func (c *AWSLBReadvertiserController) createTwoWayEndpointMergePatch(endpoint *c
 		return nil, fmt.Errorf("failed to patch bytes")
 	}
 
-	_, err = c.client.CoreV1().Endpoints(metav1.NamespaceDefault).Patch(endpoint.Name, types.StrategicMergePatchType, patchBytes)
+	_, err = c.client.CoreV1().Endpoints(metav1.NamespaceDefault).Patch(context.TODO(), endpoint.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
```
Comment (BeckerMax) [inline controller/endpoints.go:83]: Reviewers please provide feedback about the right usage of the context

File: controller/endpoint_suite_test.go [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package controller
```
Comment (rfranzke) [inline controller/endpoint_suite_test.go:1]: license header missing

File: go.mod [hunk 3]
```diff
@@ -0,0 +1,27 @@
+module github.com/gardener/aws-lb-readvertiser
+
+go 1.14
```
Comment (rfranzke) [inline go.mod:3]:
```suggestion
go 1.15
```

Can you also update this in the `Dockerfile`?

File: go.mod [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+	github.com/pkg/errors v0.9.1 // indirect
+	github.com/sirupsen/logrus v1.6.0
+	github.com/stretchr/testify v1.5.1 // indirect
+	golang.org/x/oauth2 v0.0.0-20191202225959-858c2ad4c8b6 // indirect
+	golang.org/x/time v0.0.0-20191024005414-555d28b269f0 // indirect
+	google.golang.org/appengine v1.6.5 // indirect
+	gopkg.in/check.v1 v1.0.0-20200902074654-038fdea0a05b // indirect
+	k8s.io/api v0.18.8
+	k8s.io/apimachinery v0.18.8
+	k8s.io/client-go v0.18.2
```
Comment (rfranzke) [inline go.mod:24]:
```suggestion
	k8s.io/client-go v0.18.8
```

File: main.go [hunk 5]
```diff
@@ -53,7 +53,7 @@ func (a *AWSReadvertiserOptions) validateFlags() error {
 	}
 
 	if a.refreshPeriod == 0 {
-		log.Infof("The referesh period was not set, using default %d", a.refreshPeriod)
+		log.Infof("The refersh period was not set, using default %d", a.refreshPeriod)
```
Comment (rfranzke) [inline main.go:56]: You replaced a typo with another typo... 🤣 can you fix it?

File: controller/endpoints.go [hunk 6]
```diff
@@ -53,7 +53,7 @@ func NewAWSLBEndpointsController(client kubernetes.Interface, endpointsInformer
 	return awsLBReadvertiserController
 }
 
-func (c *AWSLBReadvertiserController) createTwoWayEndpointMergePatch(endpoint *corev1.Endpoints, dnsRecords []string) (*corev1.EndpointSubset, error) {
+func (c *AWSLBReadvertiserController) applyTwoWayEndpointMergePatch(endpoint *corev1.Endpoints, dnsRecords []string) (*corev1.EndpointSubset, error) {
```
Comment (rfranzke) [inline controller/endpoints.go:56]:
```suggestion
func (c *AWSLBReadvertiserController) applyTwoWayEndpointMergePatch(ctx context.Context, endpoint *corev1.Endpoints, dnsRecords []string) (*corev1.EndpointSubset, error) {
```

File: controller/endpoints.go (hunk 1 above)
Comment (rfranzke) [reply controller/endpoints.go:83]:
```suggestion
	_, err = c.client.CoreV1().Endpoints(metav1.NamespaceDefault).Patch(ctx, endpoint.Name, types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{})
```

File: controller/endpoints.go [hunk 7]
```diff
@@ -125,12 +125,12 @@ func (c *AWSLBReadvertiserController) Run(ctx context.Context, refreshTicker *ti
 					return fmt.Errorf("%s warning: could not resolve the DNS name of the elb: %v", time.Now(), err)
 				}
 
-				endpoint, err = c.client.CoreV1().Endpoints(metav1.NamespaceDefault).Create(&corev1.Endpoints{
+				endpoint, err = c.client.CoreV1().Endpoints(metav1.NamespaceDefault).Create(context.TODO(), &corev1.Endpoints{
```
Comment (rfranzke) [inline controller/endpoints.go:128]:
```suggestion
				endpoint, err = c.client.CoreV1().Endpoints(metav1.NamespaceDefault).Create(ctx, &corev1.Endpoints{
```

File: controller/endpoints.go [hunk 8]
```diff
@@ -155,7 +155,7 @@ func (c *AWSLBReadvertiserController) Run(ctx context.Context, refreshTicker *ti
 			// handle the case where endpoint exists but has no subsets
 			if len(endpoint.Subsets) == 0 {
 				log.Infof("Found empty %s endpoint, adding correct LB IPs", endpointName)
-				endpoints, err := c.createTwoWayEndpointMergePatch(endpoint, dnsRecords)
+				endpoints, err := c.applyTwoWayEndpointMergePatch(endpoint, dnsRecords)
```
Comment (rfranzke) [inline controller/endpoints.go:158]:
```suggestion
				endpoints, err := c.applyTwoWayEndpointMergePatch(ctx, endpoint, dnsRecords)
```

File: controller/endpoints.go [hunk 9]
```diff
@@ -182,7 +182,7 @@ func (c *AWSLBReadvertiserController) Run(ctx context.Context, refreshTicker *ti
 
 				log.Info("ELB records changed, reconciling cluster endpoint to match")
 
-				endpoints, err := c.createTwoWayEndpointMergePatch(endpoint, dnsRecords)
+				endpoints, err := c.applyTwoWayEndpointMergePatch(endpoint, dnsRecords)
```
Comment (rfranzke) [inline controller/endpoints.go:185]:
```suggestion
				endpoints, err := c.applyTwoWayEndpointMergePatch(ctx, endpoint, dnsRecords)
```

**Review by rfranzke — CHANGES_REQUESTED** [review]
/assign

**Comment by rfranzke** [comment]
Can you please rebase the PR to the latest `master`?

**Comment by rfranzke** [comment]
The build step is failing with vendor issues, I guess, can you check?

File: go.mod (hunk 3 above)
Comment (BeckerMax) [reply go.mod]: Do you mean the image in the pipeline_defintions?

Comment (rfranzke) [reply go.mod]: Ah, this repository is still setup in a different/old way. Then I suggest to just update the alpine version here: https://github.com/gardener/aws-lb-readvertiser/blob/master/Dockerfile#L5

**Comment by BeckerMax** [comment]
@rfranzke everything green now. Please check again.

**Review by rfranzke — APPROVED** [review]
/lgtm

**Review by vpnachev — APPROVED** [review]
/lgtm

**Pull request merged**